### ISSUE TITLE
node/string_decoder: GenericDecoder doesn't handle incomplete chars correctly

### ISSUE THREAD

**Issue by kt3k** [denoland/deno_std#2859]
If the multi-byte chars are cut in the middle in the input, GenericDecoder (in node/string_decoder.ts) doesn't handle them correctly. The following script prints invalid chars.

```js
import { StringDecoder } from "./node/string_decoder.ts";
const text = "あいう";
const bytes = new TextEncoder().encode(text);
const dec = new StringDecoder();
let str = "";
str += dec.write(bytes.slice(0, 4));
str += dec.write(bytes.slice(4));
console.log(str);
```

For reference, the following node.js script prints correct original text.

```js
import { StringDecoder } from "string_decoder";
const text = "あいう";
const bytes = new TextEncoder().encode(text);
const dec = new StringDecoder();
let str = "";
str += dec.write(bytes.slice(0, 4));
str += dec.write(bytes.slice(4));
console.log(str);
```

**Comment by rush1506** [comment]
Has anyone works on this yet? If not, can I take this up as my first issue? :)

**Comment by cjihrig** [comment]
@rush1506 go for it.

**Comment by kt3k** [comment]
fixed in #2897
